internal/powerplatform/services/data_record/resource_data_record.go

```diff
@@ -0,0 +1,337 @@
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func convertColumnsToState(ctx context.Context, apiClient *DataRecordClient, currentState *DataRecordResourceModel, new_environment_id string, new_table_logical_name string, new_columns map[string]interface{}) *DataRecordResourceModel {
+	var objectType = map[string]attr.Type{
+		"entity_logical_name": types.StringType,
+		"data_record_id":      types.StringType,
+	}
+
+	var old_columns map[string]interface{}
+	jsonColumns, _ := json.Marshal(currentState.Columns.String())
+	unquotedJsonColumns, _ := strconv.Unquote(string(jsonColumns))
+	json.Unmarshal([]byte(unquotedJsonColumns), &old_columns)
+
+	attributeTypes := make(map[string]attr.Type)
+	attributes := make(map[string]attr.Value)
+
+	for key, value := range old_columns {
+		switch value.(type) {
+		case bool:
+			v, ok := new_columns[key].(bool)
+			if ok {
+				attributeTypes[key] = types.BoolType
+				attributes[key] = types.BoolValue(v)
+			}
+		case int64:
+			v, ok := new_columns[key].(int64)
+			if ok {
+				attributeTypes[key] = types.Int64Type
+				attributes[key] = types.Int64Value(v)
+			}
+		case float64:
+			v, ok := new_columns[key].(float64)
+			if ok {
+				attributeTypes[key] = types.Float64Type
+				attributes[key] = types.Float64Value(v)
+			}
+		case string:
+			v, ok := new_columns[key].(string)
+			if ok {
+				attributeTypes[key] = types.StringType
+				attributes[key] = types.StringValue(v)
+			}
+		case map[string]interface{}:
+			v, ok := new_columns[fmt.Sprintf("_%s_value", key)].(string)
+			if ok {
+				entityLogicalName, _ := apiClient.GetEntityRelationDefiitionInfo(ctx, new_environment_id, new_table_logical_name, key)
+				dataRecordId := v
+
+				nestedObjectType := types.ObjectType{
+					AttrTypes: objectType,
+				}
+				nestedObjectValue, _ := types.ObjectValue(
+					objectType,
+					map[string]attr.Value{
+						"entity_logical_name": types.StringValue(entityLogicalName),
+						"data_record_id":      types.StringValue(dataRecordId),
+					},
+				)
+
+				attributeTypes[key] = nestedObjectType
+				attributes[key] = nestedObjectValue
+			}
+		case []interface{}:
+			var listTypes []attr.Type
+			var listValues []attr.Value
+			tupleElementType := types.ObjectType{
+				AttrTypes: objectType,
+			}
+
+			relationMap, _ := apiClient.GetRelationData(ctx, currentState.Id.ValueString(), new_environment_id, new_table_logical_name, key)
+			entityLogicalName, primaryIdFieldName := apiClient.GetEntityRelationDefiitionInfo(ctx, new_environment_id, new_table_logical_name, key)
+
+			for _, value := range relationMap {
+				item := value.(map[string]interface{})
+
+				field := item[primaryIdFieldName]
+				if field == nil {
+					field = item[fmt.Sprintf("_%s_value", primaryIdFieldName)]
+				}
+
+				v, _ := types.ObjectValue(objectType, map[string]attr.Value{
+					"entity_logical_name": types.StringValue(entityLogicalName),
```

Comment:
I think we should be consistent with `table_logical_name`.   Dataverse rebranded from "entities" to "tables" a few years ago.  We seem to be inconsistent and sometimes use `table_logical_name` and sometimes call it `entity_logical_name`.  I think we should always use "table" over "entity".  @mawasile thoughts?

Reply:
updated 👍